crates/valence_entity/src/active_status_effects.rs

```diff
@@ -0,0 +1,522 @@
+                effects.append(&mut remaining_effects);
+                true
+            } else if active_status_effect.remaining_duration() > duration
+                && active_status_effect.amplifier() < amplifier
+            {
+                // if its duration is longer and its amplifier is lower, insert
+                // the new effect before it.
+                effects.insert(index, effect);
+                true
+            } else {
+                // if its duration is longer and its amplifier is higher, do
+                // nothing.
+                false
+            }
+        } else {
+            // Found no effect with an equal or lower amplifier.
+            // This means all effects have a higher amplifier or the vec is
+            // empty.
+
+            if let Some(last) = effects.last() {
+                // There is at least one effect with a higher amplifier.
+                if last.remaining_duration() < effect.remaining_duration() {
+                    // if its duration is shorter, we can insert it at the end.
+                    effects.push(effect);
+                    true
+                } else {
+                    // if its duration is longer, do nothing.
+                    false
+                }
+            } else {
+                // The vec is empty.
+                effects.push(effect);
+                true
+            }
+        }
+    }
+
+    /// Replaces an effect.
+    fn replace_effect(&mut self, effect: ActiveStatusEffect) {
+        self.current_effects
+            .insert(effect.status_effect(), vec![effect]);
+    }
+
+    /// Removes an effect.
+    fn remove_effect(&mut self, effect: StatusEffect) {
+        self.current_effects.remove(&effect);
+    }
+
+    /// Removes all effects.
+    fn remove_all_effects(&mut self) {
+        self.current_effects.clear();
+    }
+
+    /// Removes the strongest effect of the given type, i.e., the first effect
+    fn remove_strongest_effect(&mut self, effect: StatusEffect) {
+        if let Some(effects) = self.current_effects.get_mut(&effect) {
+            effects.remove(0);
+        }
+    }
+
+    /// **For internal use only.**
+    ///
+    /// Increments the active tick of all effects by a tick.
+    pub fn increment_active_ticks(&mut self) {
+        for effects in self.current_effects.values_mut() {
+            for effect in effects.iter_mut() {
+                effect.increment_active_ticks();
+
+                if effect.expired() {
+                    self.changes
+                        .push(StatusEffectChange::Expire(effect.status_effect()));
+                }
+            }
+        }
+    }
+
+    /// **For internal use only.**
+    ///
+    /// Applies all the changes.
+    ///
+    /// Returns a [`IndexMap`] of [`StatusEffect`]s that were updated or removed
+    /// and their previous values.
+    pub fn apply_changes(&mut self) -> IndexMap<StatusEffect, Option<ActiveStatusEffect>> {
```

Comment:
Since this is intended to be an internal api:
```suggestion
    #[doc(hidden)]
    pub fn apply_changes(&mut self) -> IndexMap<StatusEffect, Option<ActiveStatusEffect>> {
```